### PR TITLE
[HYDRATOR-552] Adding Sampling plugin

### PR DESCRIPTION
**Pull request by romy-khetan**

### CODE REVIEW

File: docs/Sampling-batchaggregator.md [hunk 1]
```diff
@@ -0,0 +1,75 @@
+# Sampling Aggregate
+
+
+Description
+-----------
+Sampling a large dataset flowing through this plugin to pull random records. Supports two types of sampling
+i.e, Systematic Sampling and Reservoir Sampling.
+
+
+Use Case
```
Comment (russorat) [inline docs/Sampling-batchaggregator.md]: remove this section

File: docs/Sampling-batchaggregator.md [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+--------
+The plugin is used when you want to randomly sample a large dataset based on different sampling types.
+
+
+Properties
+----------
+**sampleSize:** The number of records that needs to be sampled from the input records.
+
+**samplePercentage:** The percentage of records that needs to be sampled from the input records. Either of
+'samplePercentage' or 'sampleSize' needs to be mentioned.
```
Comment (russorat) [inline docs/Sampling-batchaggregator.md]: needs to be mentioned -> should be specified for this plugin.

File: docs/Sampling-batchaggregator.md [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+
+**samplingType:** Type of the Sampling algorithm that needs to be used to sample the data.
+For example: Systematic or Reservoir
```
Comment (russorat) [inline docs/Sampling-batchaggregator.md]: Type of the Sampling algorithm that should to be used to sample the data. This can be either Systematic or Reservoir.

File: docs/Sampling-batchaggregator.md [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+
+**overSamplingPercentage:** The percentage of additional records that needs to be included in addition to the input
+sample size to account for oversampling to be used in Systematic Sampling.
```
Comment (russorat) [inline docs/Sampling-batchaggregator.md]: sample size to account for oversampling. Required for Systematic Sampling.

File: docs/Sampling-batchaggregator.md [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+----------
+**sampleSize:** The number of records that needs to be sampled from the input records.
+
+**samplePercentage:** The percentage of records that needs to be sampled from the input records. Either of
+'samplePercentage' or 'sampleSize' needs to be mentioned.
+
+**samplingType:** Type of the Sampling algorithm that needs to be used to sample the data.
+For example: Systematic or Reservoir
+
+**overSamplingPercentage:** The percentage of additional records that needs to be included in addition to the input
```
Comment (russorat) [inline docs/Sampling-batchaggregator.md]: that needs to be included -> that should be included

File: docs/Sampling-batchaggregator.md [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+sample size to account for oversampling to be used in Systematic Sampling.
+
+**random:** Random float value between 0 and 1 to be used in Systematic Sampling. If not provided, plugin will
+internally generate random value.
+
+**totalRecords:** Total number of input records to be used in Systematic Sampling.
```
Comment (russorat) [inline docs/Sampling-batchaggregator.md]: Total number of input records for Systematic Sampling.

File: src/main/java/co/cask/hydrator/plugin/batch/aggreagtor/aggregator/Sampling.java [hunk 7]
```diff
@@ -0,0 +1,238 @@
...
+import java.util.List;
+import java.util.Random;
+import javax.annotation.Nullable;
+
+/**
+ * Sampling plugin to sample random data from large dataset flowing through the plugin.
+ */
+@Plugin(type = BatchAggregator.PLUGIN_TYPE)
+@Name("Sampling")
+@Description("Sampling a large dataset flowing through this plugin to pull random records.")
```
Comment (russorat) [inline src/main/java/co/cask/hydrator/plugin/batch/aggreagtor/aggregator/Sampling.java]: Sample a large dataset allowing only a subset of records through to the next stage.

File: src/main/java/co/cask/hydrator/plugin/batch/aggreagtor/aggregator/Sampling.java [hunk 8]
```diff
@@ -0,0 +1,238 @@
...
+
+    private SamplingConfig config;
+
+    public Sampling(SamplingConfig config) {
+        this.config = config;
+    }
+
+    @Override
+    public void prepareRun(BatchAggregatorContext context) throws Exception {
+        context.setNumPartitions(1);
```
Comment (russorat) [inline src/main/java/co/cask/hydrator/plugin/batch/aggreagtor/aggregator/Sampling.java]: Why is this needed?

File: src/main/java/co/cask/hydrator/plugin/batch/aggreagtor/aggregator/Sampling.java [hunk 9]
```diff
@@ -0,0 +1,238 @@
...
+
+    @Override
+    public void configurePipeline(PipelineConfigurer pipelineConfigurer) {
+        config.validate();
+        pipelineConfigurer.getStageConfigurer()
+                .setOutputSchema(pipelineConfigurer.getStageConfigurer().getInputSchema());
+    }
+
+    @Override
+    public void groupBy(StructuredRecord record, Emitter<String> emitter) throws Exception {
```
Comment (russorat) [inline src/main/java/co/cask/hydrator/plugin/batch/aggreagtor/aggregator/Sampling.java:65]: is this needed?

File: src/main/java/co/cask/hydrator/plugin/batch/aggreagtor/aggregator/Sampling.java [hunk 10]
```diff
@@ -0,0 +1,238 @@
...
+    }
+
+    /**
+     * Config for Sampling Plugin.
+     */
+    public static class SamplingConfig extends PluginConfig {
+
+        @Nullable
+        @Description("The number of records that needs to be sampled from the input records.")
+        private Integer sampleSize;
```
Comment (russorat) [inline src/main/java/co/cask/hydrator/plugin/batch/aggreagtor/aggregator/Sampling.java]: can this be macro enabled?

File: src/main/java/co/cask/hydrator/plugin/batch/aggreagtor/aggregator/Sampling.java [hunk 11]
```diff
@@ -0,0 +1,238 @@
...
+
+        @Nullable
+        @Description("The percenatage of records that needs to be sampled from the input records.")
+        private Float samplePercentage;
```
Comment (russorat) [inline src/main/java/co/cask/hydrator/plugin/batch/aggreagtor/aggregator/Sampling.java]: can this be macro enabled?

File: src/main/java/co/cask/hydrator/plugin/batch/aggreagtor/aggregator/Sampling.java [hunk 12]
```diff
@@ -0,0 +1,238 @@
...
+
+        @Description("Type of the Sampling algorithm that needs to be used to sample the data.")
+        private String samplingType;
+
+        @Nullable
+        @Description("The percenatage of additional records that needs to be included in addition to the input " +
+                "sample size to account for oversampling to be used in Systematic Sampling.")
+        private Float overSamplingPercentage;
```
Comment (russorat) [inline src/main/java/co/cask/hydrator/plugin/batch/aggreagtor/aggregator/Sampling.java]: can this be macro enabled?

File: src/main/java/co/cask/hydrator/plugin/batch/aggreagtor/aggregator/Sampling.java [hunk 13]
```diff
@@ -0,0 +1,238 @@
...
+
+        @Nullable
+        @Description("Random float value between 0 and 1 to be used in Systematic Sampling.")
+        private Float random;
```
Comment (russorat) [inline src/main/java/co/cask/hydrator/plugin/batch/aggreagtor/aggregator/Sampling.java]: can this be macro enabled?

File: src/main/java/co/cask/hydrator/plugin/batch/aggreagtor/aggregator/Sampling.java [hunk 14]
```diff
@@ -0,0 +1,238 @@
...
+
+        @Nullable
+        @Description("Total number od input records.")
+        private Integer totalRecords;
```
Comment (russorat) [inline src/main/java/co/cask/hydrator/plugin/batch/aggreagtor/aggregator/Sampling.java]: can this be macro enabled?

File: src/main/java/co/cask/hydrator/plugin/batch/aggreagtor/aggregator/Sampling.java [hunk 15]
```diff
@@ -0,0 +1,238 @@
...
+
+        public void validate() {
+            if (sampleSize == null && samplePercentage == null) {
+                throw new IllegalArgumentException("Please provide Sample size or Sample Percentage values.");
+            }
+
+            if (samplingType.equalsIgnoreCase(TYPE.SYSTEMATIC.toString()) && totalRecords == null) {
+                throw new IllegalArgumentException("Please provide value for 'Total Records' when selecting sampling " +
+                        "type as 'Systematic'.");
+            }
```
Comment (russorat) [inline src/main/java/co/cask/hydrator/plugin/batch/aggreagtor/aggregator/Sampling.java:240]: do you want to check for things like ranges here? like, if i enter a random number of 47?

File: src/main/java/co/cask/hydrator/plugin/batch/aggreagtor/aggregator/Sampling.java (hunk 9 above)
Comment (romy-khetan) [reply src/main/java/co/cask/hydrator/plugin/batch/aggreagtor/aggregator/Sampling.java:65]: To group the data for sampling algorithms defined in aggregate(), this function is needed. Please suggest, if i am missing anything.

Comment (russorat) [reply src/main/java/co/cask/hydrator/plugin/batch/aggreagtor/aggregator/Sampling.java:65]: all good

File: src/main/java/co/cask/hydrator/plugin/batch/aggreagtor/aggregator/Sampling.java (hunk 15 above)
Comment (russorat) [reply src/main/java/co/cask/hydrator/plugin/batch/aggreagtor/aggregator/Sampling.java:240]: ranges added

File: pom.xml [hunk 16]
```diff
@@ -0,0 +1,395 @@
...
+  License for the specific language governing permissions and limitations under
+  the License.
+  -->
+<project xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns="http://maven.apache.org/POM/4.0.0"
+         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+
+  <parent>
+    <artifactId>hydrator-plugins</artifactId>
+    <groupId>co.cask.hydrator</groupId>
+    <version>1.6.0-SNAPSHOT</version>
```
Comment (russorat) [inline pom.xml]: 1.6.0 isn't out yet. this should be 1.5.1

**Comment by russorat** [comment]
fails checkstyle:
{code}
[ERROR] /Users/russellsavage/workspace/sampling-aggregator/src/main/java/co/cask/hydrator/plugin/batch/aggreagtor/aggregator/Sampling.java:239:15: 'if' is not followed by whitespace. [WhitespaceAround]
[ERROR] /Users/russellsavage/workspace/sampling-aggregator/src/main/java/co/cask/hydrator/plugin/batch/aggreagtor/aggregator/Sampling.java:244:15: 'if' is not followed by whitespace. [WhitespaceAround]
{code}

this means you didn't actually build the project before checking in your latest changes.

File: docs/Sampling-batchaggregator.md [hunk 17]
```diff
@@ -0,0 +1,70 @@
...
+
+Description
+-----------
+Sampling a large dataset flowing through this plugin to pull random records. Supports two types of sampling
+i.e, Systematic Sampling and Reservoir Sampling.
+
+
+Properties
+----------
+**sampleSize:** The number of records that needs to be sampled from the input records.
```
Comment (russorat) [inline docs/Sampling-batchaggregator.md]: please add Either of 'samplePercentage' or 'sampleSize' should be specified for this plugin.

File: src/main/java/co/cask/hydrator/plugin/batch/aggreagtor/aggregator/Sampling.java [hunk 18]
```diff
@@ -0,0 +1,250 @@
...
+        return Schema.recordOf("schema", fields);
+    }
+
+    /**
+     * Config for Sampling Plugin.
+     */
+    public static class SamplingConfig extends PluginConfig {
+
+        @Nullable
+        @Description("The number of records that needs to be sampled from the input records.")
```
Comment (russorat) [inline src/main/java/co/cask/hydrator/plugin/batch/aggreagtor/aggregator/Sampling.java]: please add: Either of 'samplePercentage' or 'sampleSize' should be specified for this plugin.

File: src/main/java/co/cask/hydrator/plugin/batch/aggreagtor/aggregator/Sampling.java [hunk 19]
```diff
@@ -0,0 +1,250 @@
...
+        @Macro
+        private Integer sampleSize;
+
+        @Nullable
+        @Description("The percenatage of records that needs to be sampled from the input records.")
```
Comment (russorat) [inline src/main/java/co/cask/hydrator/plugin/batch/aggreagtor/aggregator/Sampling.java]: please add: Either of 'samplePercentage' or 'sampleSize' should be specified for this plugin.

File: src/main/java/co/cask/hydrator/plugin/batch/aggreagtor/aggregator/Sampling.java [hunk 20]
```diff
@@ -0,0 +1,250 @@
...
+        private String samplingType;
+
+        @Nullable
+        @Description("The percentage of additional records that should be included in addition to the input " +
+                "sample size to account for oversampling. Required for Systematic Sampling.")
+        @Macro
+        private Float overSamplingPercentage;
+
+        @Nullable
+        @Description("Random float value between 0 and 1 to be used in Systematic Sampling.")
```
Comment (russorat) [inline src/main/java/co/cask/hydrator/plugin/batch/aggreagtor/aggregator/Sampling.java]: add: If not provided, plugin will internally generate random value.

**Comment by romy-khetan** [comment]
@russorat I have updated the code and fixed checkstyle issues. Please review.

**Comment by russorat** [comment]
LGTM
